### PR TITLE
An improved version of Pull Request #72

### PR DESCRIPTION
**Pull request by lyakh**
Please, hit this hard! Review and test it thoroughly because I couldn't do that... I might be very much off track here. This is based solely on studying the code and the code isn't very simple :-)

### CODE REVIEW

**Comment by lgirdwood** [comment]
@lyakh can you change PR subject  line.

@ZhendanYang or @keqiaozhang can you test before is merged. This is a kernel crash on modules unload. @zhigang-wu can you provide bug number.

**Comment by jocelyn-li** [comment]
#73 #68 #33 #46 are related bugs for test @ZhendanYang or @keqiaozhang

**Comment by ghost** [comment]
I think the snd_soc_dapm_add_path() function's return value also need to be checked.
and the key point of this issue is not covered in this PR:
the default path is reached, what should we do?
the panic is caused by we goto the default path in snd_soc_dapm_new_widgets().
the pointer in w->kcontrols[] is null. we try to free it in topology stage. oops hits!

**Comment by lgirdwood** [comment]
@zhigang-wu best to place your review comments inline by clicking on commits, this way its easier to follow what you mean.

File: sound/soc/soc-dapm.c [hunk 1]
```diff
@@ -3061,33 +3071,39 @@ int snd_soc_dapm_new_widgets(struct snd_soc_card *card)
...
 		case snd_soc_dapm_out_drv:
-			dapm_new_pga(w);
+			ret = dapm_new_pga(w);
 			break;
 		case snd_soc_dapm_dai_link:
-			dapm_new_dai_link(w);
+			ret = dapm_new_dai_link(w);
 			break;
 		default:
 			break;
```
Comment (ghost) [inline sound/soc/soc-dapm.c:3098]: the root cause of the issue is:
One of the w->id in the link list is not processed here, it goes to the default case.
we have to handle this.

Comment (lgirdwood) [reply sound/soc/soc-dapm.c:3098]: @lyakh it seems to be failing checkpatch

**Comment by ZhendanYang** [comment]
@lgirdwood @Jocelyn-Li 
I tested this PR for issue #33 , #46 , #68 , #73 and the test result is following below:
#33 , this issue can still reproduce with sof-byt-rt5651.tplg, but can't reproduce with sof-byt-nocodec.tplg.
#46 , can't reproduce. I ran 100 times uploading and reloading driver and no kernel panic issue. But it may be blocked by #124 . 
#68 , this issue still exists, I think it may be duplicated with #33 .
#73 , can't reproduce.  I ran 100 times uploading and reloading driver and no kernel panic issue. But it may be blocked by #124 .

File: sound/soc/soc-dapm.c (hunk 1 above)
Comment (lyakh) [reply sound/soc/soc-dapm.c:3098]: Right, 2 errors are about the opening brace after a loop statement "list_for_each_..." is on a separate line. I saw it, I know it isn't the kernel style, but I used that style to preserve the original file formatting. That should be a separate patch to fix that. Perhaps before this one to then not have to add more such formatting issues. Or I can just fix this patch and then a separate patch should fix remaining old locations. As for "over 80 characters" warnings - yeah, 89 characters is a bit on a high end, but I see even a 92-char line in that file already, and it isn't one of printk() strings:
```
static int soc_dapm_mux_update_power(struct snd_soc_card *card,
				struct snd_kcontrol *kcontrol, int mux, struct soc_enum *e)
```
but I can make those shorter, if needed.
I think I'll make a new version with braces and long lines fixed - if those are the only problems with this patch.

**Pull request closed without merging**